include/yaclib/coro/detail/await_on_awaiter.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <yaclib/async/future.hpp>
+#include <yaclib/coro/coro.hpp>
+#include <yaclib/lazy/task.hpp>
+#include <yaclib/util/detail/atomic_counter.hpp>
+#include <yaclib/util/detail/unique_counter.hpp>
+#include <yaclib/util/type_traits.hpp>
+
+namespace yaclib::detail {
+
+template <bool Single>
+class AwaitOnEvent final
+  : public InlineCore,
+    public std::conditional_t<Single, OneCounter<NopeBase, NopeDeleter>, AtomicCounter<NopeBase, NopeDeleter>> {
+  using Base = std::conditional_t<Single, OneCounter<NopeBase, NopeDeleter>, AtomicCounter<NopeBase, NopeDeleter>>;
+
+ public:
+  AwaitOnEvent(std::size_t n, IExecutor& e) : Base{n}, _executor{e} {
+  }
+
+  void SetJob(BaseCore* core) {
+    _job = core;
+  }
+
+ private:
+  template <bool SymmetricTransfer>
+  [[nodiscard]] YACLIB_INLINE auto Impl(InlineCore& caller) noexcept {
+    if constexpr (Single) {
+      _executor.Submit(*_job);
+    } else {
+      if (this->SubEqual(1)) {
+        YACLIB_ASSERT(_job != nullptr);
+        _executor.Submit(*_job);
+      }
+    }
+    return Noop<SymmetricTransfer>();
+  }
+  [[nodiscard]] InlineCore* Here(InlineCore& caller) noexcept final {
+    return Impl<false>(caller);
+  }
+#if YACLIB_SYMMETRIC_TRANSFER != 0
+  [[nodiscard]] yaclib_std::coroutine_handle<> Next(InlineCore& caller) noexcept final {
+    return Impl<true>(caller);
+  }
+#endif
+  BaseCore* _job{nullptr};
+  IExecutor& _executor;
+};
+
+template <bool Single>
+struct [[nodiscard]] AwaitOnAwaiter final {
+  explicit AwaitOnAwaiter(IExecutor& e, BaseCore& caller) noexcept : _executor{e}, _caller{caller}, _event{1, e} {
+  }
+
+  YACLIB_INLINE bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  YACLIB_INLINE void await_suspend(yaclib_std::coroutine_handle<Promise> handle) noexcept {
+    BaseCore& core = handle.promise();
+    core._executor = &_executor;
+    _event.SetJob(&core);
+
+    if (!_caller.SetCallback(_event)) {
+      _executor.Submit(core);
+    }
+  }
+
+  constexpr void await_resume() const noexcept {
+  }
+
+ private:
+  IExecutor& _executor;
+  BaseCore& _caller;
+  AwaitOnEvent<Single> _event;
+};
+
+template <>
+class [[nodiscard]] AwaitOnAwaiter<false> final {
+ public:
+  template <typename... Cores>
+  explicit AwaitOnAwaiter(IExecutor& e, Cores&... cores) noexcept;
+
+  template <typename It>
+  explicit AwaitOnAwaiter(IExecutor& e, It it, std::size_t count) noexcept;
+
+  YACLIB_INLINE bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  YACLIB_INLINE void await_suspend(yaclib_std::coroutine_handle<Promise> handle) noexcept {
+    BaseCore& core = handle.promise();
+    core._executor = &_executor;
+    _event.SetJob(&core);
+
+    if (_event.SubEqual(1)) {
+      _executor.Submit(core);
+    }
+  }
+
+  constexpr void await_resume() const noexcept {
+  }
+
+ private:
+  IExecutor& _executor;
```

Comment:
You store executor twice, one time in event another time here.

It's unnecessary.
Please remove it from event, in event we can use executor from core if you store BaseCore& instead of Job&